lib/std/io_uring.zig

```diff
@@ -0,0 +1,841 @@
+// be re-numbered whenever new unions are interposed by the kernel. Furthermore, Zig's unions do not
+// support assignment by any union member directly as in C, without going through the union, so the
+// kernel adding new unions would also break existing Zig code.
+// We therefore use a flat struct without unions to avoid these two issues.
+// Pending https://github.com/ziglang/zig/issues/6349.
+pub const io_uring_sqe = extern struct {
+    opcode: linux.IORING_OP,
+    flags: u8 = 0,
+    ioprio: u16 = 0,
+    fd: i32 = 0,
+    off: u64 = 0,
+    addr: u64 = 0,
+    len: u32 = 0,
+    opflags: u32 = 0,
+    user_data: u64 = 0,
+    buffer: u16 = 0,
+    personality: u16 = 0,
+    splice_fd_in: i32 = 0,
+    options: [2]u64 = [2]u64{ 0, 0 }
+};
+
+// TODO Add to zig/std/os/bits/linux.zig:
+const IORING_SQ_CQ_OVERFLOW = 1 << 1;
+
+comptime {
+    assert(@sizeOf(io_uring_params) == 120);
+    assert(@sizeOf(io_uring_sqe) == 64);
+    assert(@sizeOf(io_uring_cqe) == 16);
+
+    assert(linux.IORING_OFF_SQ_RING == 0);
+    assert(linux.IORING_OFF_CQ_RING == 0x8000000);
+    assert(linux.IORING_OFF_SQES == 0x10000000);
+}
+
+pub const IO_Uring = struct {
+    fd: i32 = -1,
+    sq: SubmissionQueue,
+    cq: CompletionQueue,
+    flags: u32,
+
+    /// A friendly way to setup an io_uring, with default io_uring_params.
+    /// `entries` must be a power of two between 1 and 4096, although the kernel will make the final
+    /// call on how many entries the submission and completion queues will ultimately have,
+    /// see https://github.com/torvalds/linux/blob/v5.8/fs/io_uring.c#L8027-L8050.
+    /// Matches the interface of io_uring_queue_init() in liburing.
+    pub fn init(entries: u32, flags: u32) !IO_Uring {
+        var params = io_uring_params {
+            .sq_entries = 0,
+            .cq_entries = 0,
+            .flags = flags,
+            .sq_thread_cpu = 0,
+            .sq_thread_idle = 1000,
+            .features = 0,
+            .wq_fd = 0,
+            .resv = [_]u32{0} ** 3,
+            .sq_off = undefined,
+            .cq_off = undefined,
+        };
+        // The kernel will zero the memory of the sq_off and cq_off structs in io_uring_create(),
+        // see https://github.com/torvalds/linux/blob/v5.8/fs/io_uring.c#L7986-L8002.
+        return try IO_Uring.init_params(entries, &params);
+    }
+
+    /// A powerful way to setup an io_uring, if you want to tweak io_uring_params such as submission
+    /// queue thread cpu affinity or thread idle timeout (the kernel and our default is 1 second).
+    /// `params` is passed by reference because the kernel needs to modify the parameters.
+    /// You may only set the `flags`, `sq_thread_cpu` and `sq_thread_idle` parameters.
+    /// Every other parameter belongs to the kernel and must be zeroed.
+    /// Matches the interface of io_uring_queue_init_params() in liburing.
+    pub fn init_params(entries: u32, p: *io_uring_params) !IO_Uring {
```

Comment:
And for function it is `initParams` however im not sure if that applies to ported libraries

Reply:
https://github.com/ziglang/zig/issues/1097 again, and this is ported.

Reply:
Not exactly all ported, but we're already doing `io_uring_sqe` etc.

Reply:
By ported i mean is it supposed to be a drop-in replacement, i would guess no because of zig-style errors